bcipy/simulator/data/sampler.py

```diff
@@ -0,0 +1,228 @@
+import logging
+import random
+from abc import ABC
+from collections import defaultdict
+from pathlib import Path
+from typing import Callable, Dict, List, Tuple
+
+import numpy as np
+import pandas as pd
+
+from bcipy.simulator.data.data_engine import QueryFilter, RawDataEngine, Trial
+from bcipy.simulator.helpers.state import SimState
+
+log = logging.getLogger(__name__)
+
+
+def default_reshaper(eeg_responses: List[np.ndarray]) -> np.ndarray:
+    """Default data reshaper.
+
+    Returns
+    -------
+    ndarray with shape (channel_n, trial_n, sample_n)
+    """
+
+    channels_eeg: List[List[np.ndarray]] = [
+        [] for i in range(len(eeg_responses[0]))
+    ]
+
+    for _i, trial_channels_eeg in enumerate(eeg_responses):
+        for c_i, channel_eeg in enumerate(trial_channels_eeg):
+            channels_eeg[c_i].append(channel_eeg)
+
+    return np.array(channels_eeg)
+
+
+class Sampler(ABC):
+    """Represents a strategy for sampling signal model data from a DataEngine
+    comprised of signal data from one or more data collection sessions."""
+
+    def __init__(self, data_engine: RawDataEngine):
+        self.data_engine: RawDataEngine = data_engine
+        self.model_input_reshaper: Callable = default_reshaper
+
+    def sample(self, state: SimState) -> List[Trial]:
+        """
+        Query the data engine for a list of trials corresponding to each
+        currently displayed symbol.
+
+        Parameters
+        ----------
+            state - specifies the target symbol and current inquiry (displayed symbols).
+
+        Returns
+        -------
+            a list of Trials with an item for each symbol in the current inquiry.
+        """
+        raise NotImplementedError
+
+    def sample_data(self, state: SimState) -> np.ndarray:
+        """
+        Query for trials and reshape for signal model input according to the
+        provided reshaper.
+
+        Return:
+            ndarray of shape (n_channel, n_trial, n_sample)
+        """
+        trials = self.sample(state)
+        return self.reshaped(trials)
+
+    def sample_with_context(self,
```

Comment:
Does the **context** here include the information about where the sample is taken from (e.g. experiment #, trigger #) ?

Reply:
Yeah, that's the idea. I expect some of this will require refinement when we implement the specifics for Gaze.